### PR TITLE
MOSIP-12132

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/code/WorkflowActionCode.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package io.mosip.registration.processor.core.code;
+
+public enum WorkflowActionCode {
+	RESUME_PROCESSING, RESUME_PROCESSING_AND_REMOVE_HOTLISTED_TAG, RESUME_FROM_BEGINNING, RESUME_FROM_BEGINNING_AND_REMOVE_HOTLISTED_TAG, STOP_PROCESSING
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/code/WorkflowActionCode.java]: Can we make this one line per enum ?

File: registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/dao/RegistrationStatusDao.java [hunk 2]
```diff
@@ -256,4 +256,20 @@ public Boolean checkUinAvailabilityForRid(String rid) {
 		return registrationStatusRepositary.createQuerySelect(queryStr, params);
 	}
 
+	public List<RegistrationStatusEntity> getPausedPackets(Integer fetchSize, List<String> statusList) {
+		Map<String, Object> params = new HashMap<>();
+		String className = RegistrationStatusEntity.class.getSimpleName();
+		String alias = RegistrationStatusEntity.class.getName().toLowerCase().substring(0, 1);
+		
+
+		String queryStr = SELECT_DISTINCT + alias + FROM + className + EMPTY_STRING + alias + WHERE + alias
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/dao/RegistrationStatusDao.java:274]: In this query we have to mainly check if the statusCode = 'PAUSED' and resumeTimestamp > now() and defaultResumeAction is not null, can we remove the latestTransactionStatusCode checks?

Comment (vishwa-vyom) [reply registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/dao/RegistrationStatusDao.java:274]: Also update the unprocess packet query with status not equals to PAUSED check

File: registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/RegistrationStatusServiceImpl.java [hunk 3]
```diff
@@ -623,4 +644,6 @@ public void updateRegistrationStatusForWorkflow(InternalRegistrationStatusDto re
 				"RegistrationStatusServiceImpl::updateRegistrationStatusForWorkFlow()::exit");
 
 	}
+
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/RegistrationStatusServiceImpl.java]: Why 2 extra lines here?

File: registration-processor/registration-processor-registration-status-service-impl/src/test/java/io/mosip/registration/processor/status/dao/RegistrationStatusDaoTest.java [hunk 4]
```diff
@@ -98,6 +98,16 @@ public void testgetUnProcessedPacketCount() {
 		int count = registrationStatusDao.getUnProcessedPacketsCount(6000, 4, statusList);
 		assertEquals(count, entityCount);
 	}
+	
+	@Test
+	public void testgetgetPausedPackets() {
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-registration-status-service-impl/src/test/java/io/mosip/registration/processor/status/dao/RegistrationStatusDaoTest.java]: Why 2 get ?

File: registration-processor/registration-processor-registration-status-service-impl/src/test/java/io/mosip/registration/processor/status/service/RegistrationStatusServiceTest.java [hunk 5]
```diff
@@ -220,6 +220,18 @@ public void testGetUnProcessedPackets() {
 				statusList);
 		assertEquals("REPROCESS", dtolist.get(0).getLatestTransactionStatusCode());
 	}
+	
+	@Test
+	public void testGetPausedPackets() {
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-registration-status-service-impl/src/test/java/io/mosip/registration/processor/status/service/RegistrationStatusServiceTest.java:226]: Change test cases according to the statusCode field, and add negative test cases for exceptions

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java [hunk 6]
```diff
@@ -206,7 +214,8 @@ public void start() {
 	 */
 	@Override
 	public MessageDTO process(MessageDTO object) {
-		List<InternalRegistrationStatusDto> dtolist = null;
+		List<InternalRegistrationStatusDto> reprocessorDtolist = null;
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java]: I have changed it.

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java [hunk 7]
```diff
@@ -215,11 +224,33 @@ public MessageDTO process(MessageDTO object) {
 		regProcLogger.debug(LoggerFileConstant.SESSIONID.toString(), LoggerFileConstant.REGISTRATIONID.toString(), "",
 				"ReprocessorStage::process()::entry");
 		try {
-			dtolist = registrationStatusService.getUnProcessedPackets(fetchSize, elapseTime, reprocessCount,
+			pausedDtolist= registrationStatusService.getPausedPackets(fetchSize,
+					statusList);
+			List<String> deafaultActionIds=new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java]: I have removed it.

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java [hunk 8]
```diff
@@ -215,11 +224,33 @@ public MessageDTO process(MessageDTO object) {
...
+			pausedDtolist.forEach(s -> deafaultActionIds.add(s.getRegistrationId()));
+			Map<String,List<String>> map=new HashMap<>();
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java]: I have changed it.

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java [hunk 9]
```diff
@@ -215,11 +224,33 @@ public MessageDTO process(MessageDTO object) {
...
+			for(InternalRegistrationStatusDto dto:pausedDtolist) {
+					if(map.containsKey(dto.getDefaultResumeAction())) {
+						List<String> ids=map.get(dto.getDefaultResumeAction());
+						ids.add(dto.getRegistrationId());
+						map.put(dto.getDefaultResumeAction(), ids);
+					}
+					else if(!map.containsKey(dto.getDefaultResumeAction())) {
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java]: I have changed it.

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java [hunk 10]
```diff
@@ -215,11 +224,33 @@ public MessageDTO process(MessageDTO object) {
 		regProcLogger.debug(LoggerFileConstant.SESSIONID.toString(), LoggerFileConstant.REGISTRATIONID.toString(), "",
 				"ReprocessorStage::process()::entry");
 		try {
-			dtolist = registrationStatusService.getUnProcessedPackets(fetchSize, elapseTime, reprocessCount,
+			pausedDtolist= registrationStatusService.getPausedPackets(fetchSize,
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java]: I have changed it.

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java [hunk 11]
```diff
@@ -215,11 +224,33 @@ public MessageDTO process(MessageDTO object) {
...
+				
+			}
+			
+			reprocessorDtolist = registrationStatusService.getUnProcessedPackets(fetchSize, elapseTime, reprocessCount,
 					statusList);
 
-			if (!CollectionUtils.isEmpty(dtolist)) {
-				dtolist.forEach(dto -> {
+			if (!CollectionUtils.isEmpty(reprocessorDtolist)) {
+				reprocessorDtolist.removeIf(x ->deafaultActionIds.contains(x.getRegistrationId()));
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/main/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStage.java]: I have changed it.

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/test/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStageTest.java [hunk 12]
```diff
@@ -119,16 +127,22 @@ public void setup() throws Exception {
 	}
 
 	@Test
-	public void testProcessValid() {
+	public void testProcessValid() throws TablenotAccessibleException, PacketManagerException, ApisResourceAccessException {
 
 		List<InternalRegistrationStatusDto> dtolist = new ArrayList<>();
+		List<InternalRegistrationStatusDto> dtolist1 = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/test/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStageTest.java]: I have changed it.

File: registration-processor/core-processor/registration-processor-reprocessor-stage/src/test/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStageTest.java [hunk 13]
```diff
@@ -139,22 +153,29 @@ public void testProcessValid() {
...
 		dto = reprocessorStage.process(dto);
 		assertTrue(dto.getIsValid());
 	}
 
 	@Test
-	public void testProcessFailure() {
+	public void testProcessFailure() throws TablenotAccessibleException, PacketManagerException, ApisResourceAccessException {
 
 		List<InternalRegistrationStatusDto> dtolist = new ArrayList<>();
+		List<InternalRegistrationStatusDto> dtolist1 = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (ghost) [reply registration-processor/core-processor/registration-processor-reprocessor-stage/src/test/java/io/mosip/registration/processor/reprocessor/stage/ReprocessorStageTest.java]: I have changed it.

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/code/WorkflowActionCode.java (hunk 1 above)
Comment (ghost) [reply registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/code/WorkflowActionCode.java]: I have changed it.

File: registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/dao/RegistrationStatusDao.java (hunk 2 above)
Comment (ghost) [reply registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/dao/RegistrationStatusDao.java:274]: it should be resumeTimestamp < now() as we want packets which has been paused beyond resumetimestamp. There was no PAUSED status code, so I have added it to the enum "RegistrationStatusCode" and implemented it in the query.

File: registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/RegistrationStatusServiceImpl.java (hunk 3 above)
Comment (ghost) [reply registration-processor/registration-processor-registration-status-service-impl/src/main/java/io/mosip/registration/processor/status/service/impl/RegistrationStatusServiceImpl.java]: I have removed the white space.

File: registration-processor/registration-processor-registration-status-service-impl/src/test/java/io/mosip/registration/processor/status/dao/RegistrationStatusDaoTest.java (hunk 4 above)
Comment (ghost) [reply registration-processor/registration-processor-registration-status-service-impl/src/test/java/io/mosip/registration/processor/status/dao/RegistrationStatusDaoTest.java]: I have changed it.

**Review by vishwa-vyom — APPROVED** [review]
I am approving and merging this PR, but please take care of 2 review comments and raise one more PR

**Pull request merged**